design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,195 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to immediately propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out DaemonSets.
+
+
+## Goals
+
+-   Provide a way to propagate changes from a running VM object to its corresponding VMI immediately without interrupting the workload.
+    
+-   Maintain backwards compatibility
+    
+
+## Non Goals
+
+-   Successfully apply all the changes to the Virtual Machine Instance Object. Not all fields can be propagated to the guest immediately. Some fields should remain staged, such as those that affect the guest ABI (e.g. machine type)
+    
+-   Make LiveUpdate the default. We defer this decision.
+    
+
+## Definition of Users
+
+A user - a VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a user I want changes that I applied to the VM object to propagate to the VMI immediately if possible.
+
+As a user, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+A new field called rolloutStrategy will be introduced in the VM spec. This field will control the behavior of the VM object. Two possible values will be available:
+
+1.  Setting RolloutStrategy to `LiveUpdate` means all changes to the VM object will be immediately applied to the VMI.
+    
+
+A condition called `RestartRequired` will be added to the VM if such changes are not possible to propagate,
```

Comment:
Well, there must be a way to communicate that a desired state is not reachable without an action.
Conditions fit the bill from my point of view.
What is your concern?

Reply:
As a VM owner, I'd love to know that my VMI is not yet matching its desired state, and I'd like to know why. A condition like `RestartRequired` with details like `sockets > maxSockets; error hotplugging nic1;` would be very useful.

Reply:
> What is your concern?

I just wanted to know if this is needed for the users/clients or for an internal logic when processing the reconcile loop.
The only disadvantage that I can think of is that it is another thing to manage (e.g. if the VM is changed, the condition needs to be updated or removed accordingly).
If this is not needed for the internal logic, I would defer it to later and consider it an optional enhancement.

> As a VM owner, I'd love to know that my VMI is not yet matching its desired state, and I'd like to know why. A condition like RestartRequired with details like sockets > maxSockets; error hotplugging nic1; would be very useful.

Yes, that would be useful. However, a client could also calculate the diff in this regard and expose the gaps. The only thing it will not show is the reason it did not work (e.g. "not supported", "no permission").
But we also have logs and events that can be used to communicate things to users.

While I agree it may be nice to have it, we may (again) overdo things prematurely.
Said that, I do not have strong opinions on this, after all it does make sense.